### PR TITLE
Implement Square Pad and Resize Preprocessors

### LINKED ISSUE #2082: Implementation of a Resizing Preprocessor for Object Detector Inputs
**Is your feature request related to a problem? Please describe.**
The YOLOv3 and YOLOv5 estimators supported in PyTorch require images to be shaped accordingly in an appropriate size (e.g., 416x416). The issue with resizing images for object detector inputs is that the bounding boxes need to also be resized. A preprocessor object should be created that resizes images and can optionally resize bounding boxes in provided.

**Describe the solution you'd like**
Just like with `art.preprocesor.audio`, a new module `art.preprocessor.image` will be created first. Under this, the preprocessor will be implemented. This resizer preprocessor object will be instantiated to have a fixed resizing shape and will resize all images and bounding boxes provided to that shape. This preprocessor will be compatible with both classification and object detection with the only difference being that it will resize bounding boxes if they are provided.

**Describe alternatives you've considered**
This preprocessor can be implemented directly under `art.preprocessing`, but creating the `image` submodule will help future-proof for any additional non-EOT preproccesord that will be implemented.

**Additional context**
This will be a generic non-framework specific preprocessor that works on numpy inputs.

### LINKED ISSUE #2130: Implementation of a Square Padding Preprocessor for Object Detector Inputs
**Is your feature request related to a problem? Please describe.**
Follow-up on #2082 where the YOLOv3 and YOLOv5 estimators supported in PyTorch require images to be shaped accordingly in an appropriate size (e.g., 416x416). Prior to resizing the images and bounding boxes, they also need to be padded to a square shape in order to preserve the aspect ratio. Therefore, in addition to the resizing preprocessor, a square padding preprocessor will also be created. This will work for both classification and object detection classes where the bounding boxes are adjusted accordingly.

**Describe the solution you'd like**
In the new module `art.preprocessor.image` will, this square padding preprocessor will be implemented. This preprocessor object will be instantiated to pad all images to a square shape based on the current height and width and shift the bounding boxes by the padding. This preprocessor will be compatible with both classification and object detection with the only difference being that it will adjust bounding boxes if they are provided.

**Describe alternatives you've considered**
Similar to the in-progress resizing preprocessor, this square padding preprocessor can be implemented directly under `art.preprocessing`, but creating the `image` submodule will help future-proof for any additional non-EOT preproccesord that will be implemented.

**Additional context**
This issue will also be worked on in conjunction with #2082

### PR DESCRIPTION
**Pull request by f4str**
# Description

A new module `art.preprocessing.image` was created which will contain all image preprocessors. Under this module, new image preprocessors for square padding and resizing were created. Both of these preprocessors work for both the classification and object detection tasks. For classification, the labels are not changed. For object detection, the bounding boxes of the labels are shifted/scaled accordingly based on the padding or resizing.

The Square Pad preprocessor (`ImageSquarePad`, `ImageSquarePadPyTorch`, and `ImageSquarePadTensorFlowV2`) will pad all images to be a square shape. The user is capable of specifying the type of padding (e.g., constant or symmetric) and any additional parameters taken by the corresponding framework-specific padding function. This preprocessor can also take either a ndarray/tensor of the same image size or a list of ndarray/tensor of arbitrary sizes and still square pad accordingly.

The Resize preprocessor (`ImageResize`, `ImageResizePyTorch`, and `ImageResizeTensorFlowV2`) will resize all image to the be same user specified height and width. The user is capable of specifying the interpolation method used for resizing (e.g., bilinear or area). This preprocessor can also take either a ndarray/tensor of the same image size or a list of ndarray/tensor of arbitrary sizes and still resize all images to the same dimensions. This preprocessor will always convert the resized images into one large ndarray/tensor so this is ideal last step for the training pipeline.

For training, the Square Pad preprocessor should be followed by the Resize processor so images of arbitrary sizes can be padded to preserve their aspect ratio and then resized to be fit into a model. These preprocessors are necessary for properly training the YOLO estimators.

Fixes #2130 
Fixes #2082 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit tests for Square Pad preprocessor
- [x] Unit tests for Resize preprocessor

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/preprocessing/image/test_image_resize.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/preprocessing/image/test_image_resize.py:1]: I think a new call to pytest for tests/preprocessing/image is needed in run_tests.sh after line 39.

File: art/preprocessing/image/image_square_pad/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        super().__init__(is_fitted=True, apply_fit=apply_fit, apply_predict=apply_predict)
+        self.channels_first = channels_first
+        self.label_type = label_type
+        self.pad_mode = pad_mode
+        self.pad_kwargs = pad_kwargs if pad_kwargs is not None else {}
+        self.clip_values = clip_values
+        self.verbose = verbose
+        self._check_params()
+
+    def forward(  # type: ignore
```
Comment (beat-buesser) [inline art/preprocessing/image/image_square_pad/tensorflow.py:80]: Are the `type: ignore` comments in the `forward` and `__call__` methods because of the object detection label format?

File: tests/preprocessing/image/test_image_resize.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+def image_batch(channels_first, label_type):
+    channels = 3
+    height = 20
+    width = 16
+
+    if channels_first:
+        data_shape = (2, channels, height, width)
+    else:
+        data_shape = (2, height, width, channels)
+    x = (0.5 * np.ones((data_shape))).astype(ART_NUMPY_DTYPE)
```
Comment (beat-buesser) [inline tests/preprocessing/image/test_image_resize.py:41]: Potentially redundant parentheses:
```suggestion
    x = (0.5 * np.ones(data_shape)).astype(ART_NUMPY_DTYPE)
```

File: tests/preprocessing/image/test_image_square_pad.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+@pytest.fixture()
+def image_batch(height, width, channels_first, label_type):
+    channels = 3
+
+    if channels_first:
+        data_shape = (2, channels, height, width)
+    else:
+        data_shape = (2, height, width, channels)
+    x = (0.5 * np.ones((data_shape))).astype(ART_NUMPY_DTYPE)
```
Comment (beat-buesser) [inline tests/preprocessing/image/test_image_square_pad.py:38]: Potentially redundant parentheses:
```suggestion
    x = (0.5 * np.ones(data_shape)).astype(ART_NUMPY_DTYPE)
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @f4str Thank you very much for your pull request! I have added a few questions, what do you think?

File: art/preprocessing/image/image_square_pad/tensorflow.py (hunk 2 above)
Comment (f4str) [reply art/preprocessing/image/image_square_pad/tensorflow.py:80]: Yes, the `type: ignore` is needed due to the object detection labels having a different type. The alternative would be to modify the types of the base preprocessor class which would be too drastic of a change for now.

**Comment by f4str** [comment]
Hi @beat-buesser thank you for the comments! I've addressed the review comments and have made the corresponding changes. Feel free to let me know if there any other issues.

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for contributing image preprocessors to ART!
